deploy/ciao-deploy-server/main.go

```diff
@@ -0,0 +1,476 @@
+	}
+
+	err = deploy.SudoCopyFile(systemCaCertPath, caCertFile.Name())
+	if err != nil {
+		return "", "", errors.Wrap(err, "Error copying CA certificate to system location")
+	}
+
+	fmt.Printf("Scheduler certificate created in: %s\n", systemCertPath)
+	fmt.Printf("CA certificate installed in: %s\n", systemCaCertPath)
+	return systemCertPath, systemCaCertPath, nil
+}
+
+func createConfiguration() (string, error) {
+	config := &payloads.Configure{}
+	config.InitDefaults()
+	config.Configure.Scheduler.ConfigStorageURI = "/etc/ciao/configuration.yaml"
+
+	config.Configure.Storage.CephID = cephID
+
+	config.Configure.Controller.HTTPSCACert = httpsCaCertPath
+	config.Configure.Controller.HTTPSKey = httpsCertPath
+
+	config.Configure.Controller.IdentityUser = identityUser
+	config.Configure.Controller.IdentityPassword = identityPassword
+	config.Configure.IdentityService.URL = identityURL
+
+	config.Configure.Controller.AdminPassword = adminPassword
+	config.Configure.Controller.AdminSSHKey = adminSSHKey
+
+	config.Configure.Launcher.ComputeNetwork = []string{computeNet}
+	config.Configure.Launcher.ManagementNetwork = []string{mgmtNet}
+	config.Configure.Launcher.DiskLimit = false
+	config.Configure.Launcher.MemoryLimit = false
+
+	data, err := yaml.Marshal(config)
+	if err != nil {
+		return "", errors.Wrap(err, "Error creating marshalling configuration data")
+	}
+
+	f, err := ioutil.TempFile("", "configuration.yaml")
+	if err != nil {
+		return "", errors.Wrap(err, "Error creating temporary file")
+	}
+
+	_, err = f.Write(data)
+	if err != nil {
+		return "", errors.Wrap(err, "Error writing data to temporary file")
+	}
+
+	err = f.Close()
+	if err != nil {
+		return "", errors.Wrap(err, "Error closing temporary file")
+	}
+
+	return f.Name(), nil
+}
+
+func installScheduler(caCertPath, certPath string) error {
+	fmt.Println("Installing scheduler")
+	err := deploy.SudoMakeDirectory(ciaoConfigurationDirectory)
+	if err != nil {
+		return errors.Wrap(err, "Error making configuration directory")
+	}
+
+	fmt.Println("Creating configuration file")
+	p, err := createConfiguration()
+	if err != nil {
+		return errors.Wrap(err, "Error creating configuration file")
+	}
+	defer os.Remove(p)
+
+	configPath := path.Join(ciaoConfigurationDirectory, "configuration.yaml")
+	err = deploy.SudoCopyFile(configPath, p)
+	if err != nil {
+		return errors.Wrap(err, "Error copying configuration to destination")
+	}
+
+	fmt.Printf("Configuration file created: %s\n", configPath)
+
+	err = deploy.InstallTool(toolPath("ciao-scheduler"), "ciao-scheduler", caCertPath, certPath)
+	if err != nil {
+		return errors.Wrap(err, "Error installing scheduler binary")
+	}
+
+	return nil
+}
+
+func main() {
+	flag.Parse()
+
+	var err error
+	anchorCertPath, caCertPath, err = createSchedulerCert()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error creating scheduler certificate: %s\n", err)
+		os.Exit(1)
+	}
+	if scheduler {
+		installScheduler(caCertPath, anchorCertPath)
```

Comment:
I'm hoping that that this part of the tool is idempotent for multiple runs, e.g. by generating files into a temporary location and then copying over.

In particular the client is designed to be run in two phases, the first to generate the CSR and upload. The second to download the signed certficate.

Cleaning up needs to be done carefully, if you're running the server only to offer up upgrades and then you delete the scheduler certs because of some failure to e.g. bind to the interface then the whole cluster needs to be recreated.

In this particular case I don't think failing to install scheduler should delete the certificates as they may have been already there before this invocation.